### PR TITLE
R support for nb_conda_kernels 

### PR DESCRIPTION
**Pull request by damianavila**
OK... this bring R support for the `nb_conda_kernels` extension...

In short: it recognize envs paths where R (and Python) is installed and generates a kernelspec on the fly to support that specific kernel. I have also added logo support for Python and R environments and I have prepended with a `Py-` or `R-` prefix in case the logo is not enough...

After this we need to integrate it into wakari... and that should be easy (I hope :wink:)

r_local

@mmarchetti @malev @bollwyvl

### CODE REVIEW

**Comment by bollwyvl** [comment]
Cool stuff! I'll check it out!

**Comment by bollwyvl** [comment]
Related to this: there is work going on out in ubuntu land to make more kernels available through a PPA:
https://groups.google.com/forum/#!topic/jupyter/ipnJ7yQZg6k

I'm not suggesting we need to support all of these kernels, as most of them don't even have the base language available, the but it might be a good long-term goal.
>
 * Haskell: ihaskell [1] (note errata below)
 * Julia: ijulia [2]
 * Javascript: ijavascript [3]
 * Coffeescript: jp-coffeescript [4]
 * R: irkernel [5]
 * Ruby: iruby [6]

File: nb_conda_kernels/nb_conda_kernels.py [hunk 1]
```diff
@@ -27,10 +27,10 @@ def _conda_info(self):
 
         return conda_info
 
-    def _python_executable(self):
-        """Find the python executable for each env where jupyter is installed.
+    def _all_executable(self):
```
Comment (bollwyvl) [inline nb_conda_kernels/nb_conda_kernels.py]: Haven't tried this out yet, but it cries out for an even more modular approach: we've gotten over the "do it twice" hurdle, but i feel like a little class/registration method/multipledispatch/entry_point would make this extensible to cover other kernels.

File: setup.py [hunk 2]
```diff
@@ -38,6 +38,9 @@
     if os.path.exists(pjoin(d, '__init__.py')):
         packages.append(d[len(here)+1:].replace(os.path.sep, '.'))
 
+package_data = {'nb_conda_kernels': [pjoin('logos', 'python', '*'),
```
Comment (bollwyvl) [inline setup.py]: Again, haven't run this, but: if we are installing the *actual* kernel packages, they should include their own logos, and we needn't ship them, or?

Comment (bollwyvl) [inline setup.py]: Even more so: some of the kernels ship a [kernel.js](https://github.com/gibiansky/IHaskell/blob/master/html/kernel.js)!

**Comment by bollwyvl** [comment]
@damianavila well, it _installed_ on win7/py27-32... i had some issues getting it to work with `python develop` and `nb_conda_kernels.install`, now trying to `conda build`...

File: nb_conda_kernels/nb_conda_kernels.py (hunk 1 above)
Comment (damianavila) [inline nb_conda_kernels/nb_conda_kernels.py]: I agree... this is just an intermediate step to have the R working out of the box in wakari... but my plan for this later is to abstract this part to support multiples kernels...

File: setup.py (hunk 2 above)
Comment (damianavila) [inline setup.py]: `nb_conda_kernels` is not "installing" kernelspecs for each environment... you could potentially do that and put the logos there... instead this extension create the kernelspec on the fly, so we need to tell it which logo to use for each kernel/envs... For *actual* kernels, the mechanism is just reading the kernel folder where the logos live...

**Comment by damianavila** [comment]
> @damianavila well, it installed on win7/py27-32... i had some issues getting it to work with python develop and nb_conda_kernels.install, now trying to conda build...

Thanks for testing, pretty sure there are some win issues...

File: nb_conda_kernels/nb_conda_kernels.py (hunk 1 above)
Comment (bollwyvl) [inline nb_conda_kernels/nb_conda_kernels.py]: Sounds good!

File: setup.py (hunk 2 above)
Comment (bollwyvl) [inline setup.py]: Right, but the kernel has to _be there_ in order for us to use it. If it is installed, if follows, it also has it's resource data:
```bash
~/miniconda/envs/nb_conda_kernels$ find . | grep 64x64
./lib/python3.5/site-packages/ipykernel/resources/logo-64x64.png
./lib/R/library/IRkernel/kernelspec/logo-64x64.png
./share/jupyter/kernels/ir/logo-64x64.png
```
Unfortunately, there's no reproducible way to figure that out, i'd imagine :(

**Comment by bollwyvl** [comment]
I feel pretty good about the stuff over on #6... really nothing surprising, just hadn't been tried in the Spawn of Redmond.

Back out on this issue:

## Testing
I guess one could start the notebook process, then hit the API:
```python
# after installing r-essentials (and notebook, duh)
def test_kernel_specs():
    env = os.environ.get(HOWEVER_YOU_RELIABLY_GET_ENV)
    nb = Popen(["jupyter", "notebook", "--no-browser"])
    specs = requests.get("http://localhost:8888/api/kernelspecs").json()
    # check for keys
    for spec in ["Root", "Py-" + env, "R-" + env]:
        assert spec in specs 
    nb.terminate()
```

See #5 for my issues on windows with IRKernel, though :ghost: 

## UI
As the number of things one might see in that dropdown grows (because conda environments are so easy to make :)), i already find myself getting a bit lost when picking my kernel.

Perhaps 
```
[my_env] Python 3.5
[my_other_env] Python 2.7
[my_env] R 3.2.2
[root] Python 3.5
```
or the more typographically pleasing, 
```
Python 2.7 [my_other_env]
Python 3.5 [my_env]
Python 3.5 [root]
R 3.2.2 (my_env) 
```

would be easier to read, and sort up more nicely. I guess this depends on how a user thinks about their environments: env first, or language first.

It's probably not worth adding front-end assets (and tests!) at this point, but... a little hacking of the [kernel selector](https://github.com/jupyter/notebook/blob/master/notebook/static/notebook/js/kernelselector.js) would make this really pop, putting the environment name in, say, a bootstrap badge, so that it looked more like:
screen shot 2016-01-22 at 12 46 02 pm

> prototyped with
```
$('li[id^="kernel"]').each(function(){
  var a = $(this).find("a"),
    txt = a.text().split("-");
  a.text(txt[0])
  a.append($("<label/>", {
    "class": "label label-success pull-right"
  }).text(txt.slice(1).join("-")));
})
```

## User Expectations
This is a somewhat dangerous feature!

> As a user I launch `jupyter` in my enviornment `foo`, which doesn't has `ipywidgets` installed.
I create a notebook with the `Py-bar` kernel, where `bar` is an environment where `ipywidgets` is installed.
When I try to use widgets... I get errors about ipywidgets not being installed.

This means that a user will need to maintain a fair amount of duplicated installs (luckily, conda!) for extensions they wish to use consistently... and installing them will require restarting their server.

If we decide to do something about the UI, it seems it would be worthwhile to indicate what extensions are installed/enabled/etc.

**Comment by damianavila** [comment]
Merged #6.... I will answer to your thoughts later... thanks for the feedback!!

File: conda.recipe/post-link.bat [hunk 3]
```diff
@@ -1 +1 @@
-"%PREFIX%\bin\python" -m nb_conda_kernels.install --enable --prefix="%PREFIX%"
+"%PREFIX%\python" -m nb_conda_kernels.install --enable --prefix="%PREFIX%"
```
Comment (malev) [inline conda.recipe/post-link.bat]: why not `$PYTHON`?

Comment (bollwyvl) [inline conda.recipe/post-link.bat]: Good question! I always forget that special case.

Comment (damianavila) [inline conda.recipe/post-link.bat]: You mean:

`"%PYTHON%" -m nb_conda_kernels.install --enable --prefix="%PREFIX%"`

I am OK with that...

Comment (damianavila) [inline conda.recipe/post-link.bat]: @malev ^^^

@bollwyvl looking into nbpresent, it seems you use it a reliable way... https://github.com/Anaconda-Server/nbpresent/blob/master/conda.recipe/bld.bat

Did you validated that on win for nbpresent? I mean... it is working OK?

**Comment by damianavila** [comment]
@bollwyvl 
* I will open an issue for the UI styling with static asset, it is interesting to do something on that front later... in the meantime I took the "typographically pleasing" approach
* I added some test in line what you describe
* I use now the `%PYTHON%` env variable for win pre-post scripts

>As a user I launch jupyter in my enviornment foo, which doesn't has ipywidgets installed.
I create a notebook with the Py-bar kernel, where bar is an environment where ipywidgets is installed.
When I try to use widgets... I get errors about ipywidgets not being installed.
This means that a user will need to maintain a fair amount of duplicated installs (luckily, conda!) for extensions they wish to use consistently... and installing them will require restarting their server.

I know this is an issue and it makes sense to workout a way to tell the user that they need some packages in the "run" environments or the "server" environment... we should discuss that in another issue. 

@bollwyvl @malev @mmarchetti merging this one by tonight unless there is some bad to fix before :wink:

**Comment by bollwyvl** [comment]
Ah, you hit the markupsafe error as well!
http://anaconda.org/anaconda-nb-extensions/nb_conda_kernels/builds/15/9#L310

I just downloaded and `python setup.py developed` and everything worked fine. I like the new menu: it will be great when we circle around and pretty it up a bit :)

`conda build` osx looked great. I love getting the colored text output.

I am closing in on having nbpresent testing on windows, but let me know if there is any other way I can help!

**Pull request merged**